### DISCUSSION TITLE
How to sample respective BSDFs after `scene.ray_intersect(rays)`?

### DISCUSSION THREAD

**Discussion by sunilhadap** [mitsuba-renderer/mitsuba3#1085] [Q&A]
Hi all,

I am trying to programmatically render the scene (using `cuda_ad_rgb` variant). I populate bunch of rays. Followed by a call to `scene.ray_intersect()`.

```python
rays = my_sample_rays(scene=scene, sensor=sensor, sampler=sampler)
result = scene.ray_intersect(rays) # SurfaceInteraction
```

Once I get the ray interaction result, how can I sample the BSDFs respective to each ray using the `SurfaceInteraction`? This BSDF sampling will pretty fragmented, right? How can it be done efficiently in a vectorized form?

Thanks a bunch in advance, hopefully I am not missing something obvious.

Regards

Sunil

**Comment by sunilhadap** [comment]
BTW, ideally, I would like to implement a new type of sensor, instead of this ad-hoc sampling. Would there be, or would some have, an example of how to implement plugin sensor in python?

**Reply by sunilhadap** [reply]
Answering my own question about sensor plugin, found this spherical sensor example
https://github.com/rgl-epfl/differentiable-sdf-rendering/blob/main/figures/spherical_sensor.py

**Reply by njroussel** [reply]
Hi @sunilhadap 

Have a look at how some of our [own integrators](https://github.com/mitsuba-renderer/mitsuba3/blob/master/src/python/python/ad/integrators/prb_basic.py#L104-L111) are implemented. It is also described in the API reference: [`si.bsdf()`](https://mitsuba.readthedocs.io/en/latest/src/api_reference.html#mitsuba.SurfaceInteraction3f.bsdf).

As for custom plugins, we have a tutorial on that, albeit for BSDFs: https://mitsuba.readthedocs.io/en/latest/src/others/custom_plugin.html
You simply need to implement the methods of the [`Sensor` interface](https://mitsuba.readthedocs.io/en/latest/src/api_reference.html#mitsuba.Sensor).
